File: Solutions/Neetcode/SlidingWindow/3-LongestRepeatingCharacterReplacement/solution.ts
export function characterReplacement(s: string, k: number): number {
  const windowCharFreq = new Map<string, number>();
  let maxCharFreq = 0;
  let result = 0;

  let left = 0;
  for (let right = 0; right < s.length; right++) {
    windowCharFreq.set(s[right], (windowCharFreq.get(s[right]) || 0) + 1);
    maxCharFreq = Math.max(maxCharFreq, windowCharFreq.get(s[right])!);

    let windowLength = right - left + 1;
    if (windowLength - maxCharFreq > k) {
      windowCharFreq.set(s[left], windowCharFreq.get(s[left])! - 1);
      left++;
      windowLength--;
    }

    result = Math.max(result, windowLength);
  }

  return result;
}

File: Solutions/Neetcode/SlidingWindow/3-LongestRepeatingCharacterReplacement/Solution.py
from collections import defaultdict


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        window = defaultdict(int)
        max_freq_in_window = 0
        res = 0

        l, r = 0, 0

        while r < len(s):
            window[s[r]] += 1
            max_freq_in_window = max(max_freq_in_window, window[s[r]])

            window_length = r - l + 1
            if window_length - max_freq_in_window > k:
                window[s[l]] -= 1
                l += 1
                window_length -= 1

            res = max(res, window_length)
            r += 1

        return res
